Mathlib/Probability/Moments/ComplexMGF.lean

```diff
@@ -0,0 +1,436 @@
+      = ∫ ω, X ω ^ n * rexp (t' * X ω) ∂μ := by
+    filter_upwards [isOpen_interior.eventually_mem ht] with t ht' using h_re_of_mem n t ht'
+  rw [← EventuallyEq.hasDerivAt_iff (h_re _), ← h_re_of_mem _ t ht]
+  exact (hasDerivAt_integral_pow_mul_exp (by simp [ht]) n).real_of_complex
+
+lemma hasDerivAt_iteratedDeriv_complexMGF (hz : z.re ∈ interior (integrableExpSet X μ)) (n : ℕ) :
+    HasDerivAt (iteratedDeriv n (complexMGF X μ)) μ[fun ω ↦ X ω ^ (n + 1) * cexp (z * X ω)] z := by
+  induction n generalizing z with
+  | zero => simp [hasDerivAt_complexMGF hz]
+  | succ n hn =>
+    rw [iteratedDeriv_succ]
+    have : deriv (iteratedDeriv n (complexMGF X μ))
+        =ᶠ[𝓝 z] fun z ↦ μ[fun ω ↦ X ω ^ (n + 1) * cexp (z * X ω)] := by
+      have h_mem : ∀ᶠ y in 𝓝 z, y.re ∈ interior (integrableExpSet X μ) := by
+        refine IsOpen.eventually_mem ?_ hz
+        exact isOpen_interior.preimage Complex.continuous_re
+      filter_upwards [h_mem] with y hy using HasDerivAt.deriv (hn hy)
+    rw [EventuallyEq.hasDerivAt_iff this]
+    exact hasDerivAt_integral_pow_mul_exp hz (n + 1)
+
+/-- For `z : ℂ` with `z.re ∈ interior (integrableExpSet X μ)`, the n-th derivative of the function
+`complexMGF X μ` at `z` is `μ[X ^ n * cexp (z * X)]`. -/
+lemma iteratedDeriv_complexMGF (hz : z.re ∈ interior (integrableExpSet X μ)) (n : ℕ) :
+    iteratedDeriv n (complexMGF X μ) z = μ[fun ω ↦ X ω ^ n * cexp (z * X ω)] := by
+  induction n generalizing z with
+  | zero => simp [complexMGF]
+  | succ n hn =>
+    rw [iteratedDeriv_succ]
+    exact (hasDerivAt_iteratedDeriv_complexMGF hz n).deriv
+
+end Deriv
+
+section EqOfMGF
+
+/-! We prove that if two random variables have the same `mgf`, then
+they also have the same `complexMGF`.-/
+
+variable {Ω' : Type*} {mΩ' : MeasurableSpace Ω'} {Y : Ω' → ℝ} {μ' : Measure Ω'}
+
+/-- If two random variables have the same moment generating function then they have
+the same `integrableExpSet`. -/
+lemma integrableExpSet_eq_of_mgf' (hXY : mgf X μ = mgf Y μ') (hμμ' : μ = 0 ↔ μ' = 0) :
+    integrableExpSet X μ = integrableExpSet Y μ' := by
+  ext t
+  simp only [integrableExpSet, Set.mem_setOf_eq]
+  by_cases hμ : μ = 0
+  · simp [hμ, hμμ'.mp hμ]
+  have hμ' : μ' ≠ 0 := (not_iff_not.mpr hμμ').mp hμ
+  rw [← mgf_pos_iff' hμ, ← mgf_pos_iff' hμ', hXY]
+
+/-- If two random variables have the same moment generating function then they have
+the `integrableExpSet`. -/
+lemma integrableExpSet_eq_of_mgf [IsProbabilityMeasure μ] [IsProbabilityMeasure μ']
+    (hXY : mgf X μ = mgf Y μ') :
+    integrableExpSet X μ = integrableExpSet Y μ' :=
+  integrableExpSet_eq_of_mgf' hXY <| by simp [IsProbabilityMeasure.ne_zero]
+
+/-- If two random variables have the same moment generating function then they have
+the same `complexMGF` on the vertical strip `{z | z.re ∈ interior (integrableExpSet X μ)}`. -/
+lemma eqOn_complexMGF_of_mgf' (hXY : mgf X μ = mgf Y μ') (hμμ' : μ = 0 ↔ μ' = 0) :
```

Comment:
This argument feels a little bit strange to me: if they have the same `mgf` and `integrableExpSet` has nonempty interior, then `X` and `Y` are identically distributed, and then of course they have the same `complexMGF` (and not only on the vertical strip, but everywhere). I guess we don't have the fact I've used above (same mgf => identDistrib), but it looks more fundamental to me than the current statement. Unless the current statement is a lemma on the way to a proof of this result?

Reply:
Etienne pointed this out too in one of the now hidden discussions above: the proof I have in mind for [equal mgf => equal distributions] is [equal mgf => equal complexMGF on a domain containing the vertical line going through 0 => equal characteristic functions => equal distributions].
Perhaps there is a more direct route I should use instead?

Reply:
This is a perfectly reasonable route (I would follow the same). Maybe add comments in the docstring lemma explaining this (and in particular explain that a stronger version of the lemma can be proved once this path has been taken to its end).

I wonder if it would make sense to weaken the assumptions, just assuming that the mgf coincide on a small open interval inside `integrableExpSet` of both -- which in the end should be enough to show that they are identically distributed.

Reply:
I added comments.
About the weaker assumptions: let's wait and see what would be convenient when we actually use the lemmas?

Reply:
I wonder if this lemma will be enough for the proof you're describing: if `0` is on the boundary of `integrableExpSet`, then you would rather need to know that `complexMGF` coincide on the whole strip above `integrableExpSet`, not just its interior. This follows from what you have just proved, and continuity on the whole strip (which follows from dominated convergence, say, and would in any case be good to have). This can definitely wait for another PR, though!